posthog-server/src/main/java/com/posthog/server/internal/FlagEvaluator.kt

```diff
@@ -0,0 +1,803 @@
+package com.posthog.server.internal
+
+import com.posthog.PostHogConfig
+import java.security.MessageDigest
+import java.text.Normalizer
+import java.time.Instant
+import java.time.ZoneId
+import java.time.ZonedDateTime
+import java.time.format.DateTimeFormatter
+import java.time.format.DateTimeParseException
+import java.time.temporal.ChronoUnit
+import java.util.Date
+import java.util.regex.PatternSyntaxException
+
+/**
+ * Local evaluation engine for feature flags
+ */
+internal class FlagEvaluator(
+    private val config: PostHogConfig,
+) {
+    companion object {
+        private const val LONG_SCALE = 0xFFFFFFFFFFFFFFF.toDouble()
+        private val NONE_VALUES_ALLOWED_OPERATORS = setOf("is_not")
+        private val REGEX_COMBINING_MARKS = "\\p{M}+".toRegex()
+
+        // Date formatters for parsing various date formats
+        private val DATE_FORMATTER_WITH_SPACE_TZ = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss XXX")
+        private val DATE_FORMATTER_NO_SPACE_TZ = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ssXXX")
+        private val DATE_FORMATTER_NO_TZ = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
+
+        private fun casefold(input: String): String {
+            val normalized = Normalizer.normalize(input, Normalizer.Form.NFD)
+            return REGEX_COMBINING_MARKS.replace(normalized, "").uppercase().lowercase()
+        }
+    }
+
+    private data class VariantLookupEntry(
+        val key: String,
+        val valueMin: Double,
+        val valueMax: Double,
+    )
+
+    /**
+     * Hash function for consistent rollout percentages
+     * Given the same distinct_id and key, it'll always return the same float.
+     * These floats are uniformly distributed between 0 and 1.
+     */
+    private fun hash(
+        key: String,
+        distinctId: String,
+        salt: String = "",
+    ): Double {
+        val hashKey = "$key.$distinctId$salt"
+        val digest = MessageDigest.getInstance("SHA-1")
```

Comment:
should this be a global var or do we need a fresh instance every time?

Reply:
`getInstance` is a bit of a misnomer here - it's actually creating a new, stateful digest instance. It's possible we could use a single instance, but then would need to be synchronized. In this case, I don't think it's necessarily warranted.